File: Graph.py
class Graph:
    def __init__(self):
        self.graph = {}

    def add_edge(self, u, v):
        if u not in self.graph:
            self.graph[u] = []
        self.graph[u].append(v)

    def dfs(self, start, goal, path=None):
        if path is None:
            path = []
        path.append(start)

        if start == goal:
            return path

        for neighbor in self.graph.get(start, []):
            if neighbor not in path:
                result = self.dfs(neighbor, goal, path.copy())
                if result:
                    return result

        return None

# Example usage:
g = Graph()
g.add_edge('A', 'B')
g.add_edge('A', 'C')
g.add_edge('B', 'D')
g.add_edge('C', 'D')
g.add_edge('D', 'E')

start_building = 'A'
end_building = 'E'
path = g.dfs(start_building, end_building)

if path:
    print(f"Path from {start_building} to {end_building}: {' -> '.join(path)}")
else:
    print(f"No path found from {start_building} to {end_building}")

File: Graph.java
import java.util.*;

class Graph {
    private Map<String, List<String>> graph;

    public Graph() {
        graph = new HashMap<>();
    }

    public void addEdge(String u, String v) {
        graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
    }

    public List<String> dfs(String start, String goal) {
        List<String> path = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        if (dfsHelper(start, goal, path, visited)) {
            return path;
        }
        return null;
    }

    private boolean dfsHelper(String current, String goal, List<String> path, Set<String> visited) {
        path.add(current);
        visited.add(current);

        if (current.equals(goal)) {
            return true;
        }

        for (String neighbor : graph.getOrDefault(current, Collections.emptyList())) {
            if (!visited.contains(neighbor)) {
                if (dfsHelper(neighbor, goal, path, visited)) {
                    return true;
                }
            }
        }

        path.remove(path.size() - 1);
        return false;
    }

    public static void main(String[] args) {
        Graph g = new Graph();
        g.addEdge("A", "B");
        g.addEdge("A", "C");
        g.addEdge("B", "D");
        g.addEdge("C", "D");
        g.addEdge("D", "E");

        String startBuilding = "A";
        String endBuilding = "E";
        List<String> path = g.dfs(startBuilding, endBuilding);

        if (path != null) {
            System.out.println("Path from " + startBuilding + " to " + endBuilding + ": " + String.join(" -> ", path));
        } else {
            System.out.println("No path found from " + startBuilding + " to " + endBuilding);
        }
    }
}
